### ISSUE TITLE
enabling systemd-boot-check-no-failures.service introduces systemd ordering cycles

### ISSUE THREAD

**Issue by runcom** [coreos/fedora-coreos-tracker#1298]
**Describe the bug**
enabling systemd-boot-check-no-failures.service introduces ordering cycles for zincati, afterburn-firstboot-checkin, afterburn-checkin

**Reproduction steps**
1. install FCOS
2. systemctl enable systemd-boot-check-no-failures.service
3. reboot
4. check zincati, afterburn-firstboot-checkin, afterburn-checkin, they have ordering cycle errors + not started

**Expected behavior**
they should just start I guess

**Actual behavior**
they don't start as they're removed from the transaction

**System details**
 - Bare Metal/QEMU/AWS/GCP/etc.: qemu
 - Fedora CoreOS version: latest

**Ignition config**
Please attach the Butane config or Ignition config used to provision your system. Be sure to sanitize any private data. If not using Butane to generate your Ignition config, does the Ignition config pass validation using [ignition-validate](https://coreos.github.io/ignition/getting-started/#config-validation)?

just core ssh key

**Additional information**
it's likely related to the Requires=boot-complete.target and After too - the question I do have is why this dependency on boot-complete.service if nobody even leverage that target, as it is today nothing hooks before that target to make sure it can't be reached if something goes wrong and if we do by enabling systemd-boot-check-no-failures.service everything blows

**Comment by runcom** [comment]
if we try and break the dep cycle on multi-user with:

```
[Unit]
Description=Check if Any System Units Failed
Documentation=man:systemd-boot-check-no-failures.service(8)
#After=default.target multi-user.target
Before=boot-complete.target
Conflicts=shutdown.target
Before=shutdown.target

[Service]
Type=oneshot
RemainAfterExit=yes
ExecStart=/usr/lib/systemd/systemd-boot-check-no-failures

[Install]
RequiredBy=boot-complete.target
```
this will make it work but then there's a chance that this service runs "before" multi-user is reached (right?)

**Comment by runcom** [comment]
```
Sep 15 11:30:53 localhost systemd[1]: multi-user.target: Found ordering cycle on zincati.service/start
Sep 15 11:30:53 localhost systemd[1]: multi-user.target: Found dependency on boot-complete.target/start
Sep 15 11:30:53 localhost systemd[1]: multi-user.target: Found dependency on systemd-boot-check-no-failures.service/s>
Sep 15 11:30:53 localhost systemd[1]: multi-user.target: Found dependency on multi-user.target/start
Sep 15 11:30:53 localhost systemd[1]: multi-user.target: Job zincati.service/start deleted to break ordering cycle st
```

**Comment by lucab** [comment]
I don't think this an FCOS bug, but in general a problem with the target ordering and dependencies.

Last step at https://systemd.io/AUTOMATIC_BOOT_ASSESSMENT/#how-to-adapt-this-scheme-to-other-setups says:
> To support additional components that shall only run on boot success, simply wrap them in a unit and order them after boot-complete.target, pulling it in.

So let's just do a minimal reproducer according to that:
```
# systemctl disable zincati.service afterburn-checkin.service afterburn-firstboot-checkin.service
Removed /etc/systemd/system/multi-user.target.wants/afterburn-checkin.service.
Removed /etc/systemd/system/multi-user.target.wants/afterburn-firstboot-checkin.service.
Removed /etc/systemd/system/multi-user.target.wants/zincati.service.

# cat /etc/systemd/system/all-systems-go.service 
[Unit]
Description="Order after boot-complete.target, pulling it in"
Requires=boot-complete.target
After=boot-complete.target

[Service]
Type=oneshot
ExecStart=/usr/bin/echo "Additional component that shall only run on boot success"
RemainAfterExit=yes

[Install]
WantedBy=default.target

# systemctl enable all-systems-go.service  systemd-boot-check-no-failures.service 
Created symlink /etc/systemd/system/default.target.wants/all-systems-go.service → /etc/systemd/system/all-systems-go.service.
Created symlink /etc/systemd/system/boot-complete.target.requires/systemd-boot-check-no-failures.service → /usr/lib/systemd/system/systemd-boot-check-no-failures.service.

# systemctl reboot
```

That minimal `all-systems-go.service ` example alone in conjunction with `systemd-boot-check-no-failures.service` would result in a loop:
```
systemd[1]: multi-user.target: Found ordering cycle on all-systems-go.service/start
systemd[1]: multi-user.target: Found dependency on boot-complete.target/start
systemd[1]: multi-user.target: Found dependency on systemd-boot-check-no-failures.service/start
systemd[1]: multi-user.target: Found dependency on multi-user.target/start
systemd[1]: multi-user.target: Job all-systems-go.service/start deleted to break ordering cycle starting with multi-user.target/start
```

Which indeed seems to be coming from the combination of `Before=` and `After=` in `systemd-boot-check-no-failures.service`.

I guess the real points to solve with systemd upstream here are:
 * what is the expected relative order of `boot-complete.target` and the usual default/multi-user targets?
 * what's the correct way to perform the above mentioned "order [units] after boot-complete.target", if they cannot be pulled in through the usual default/multi-user targets?

**Comment by LorbusChris** [comment]
```
[Install]
WantedBy=default.target
```
implies `Before=default.target`. If you put a `DefaultDependencies=No` on the `all-systems-go.service` Unit, it should work I htink (haven't tested it myself though).

**Comment by runcom** [comment]
@LorbusChris it does indeed work, it should still be documented by systemd to add that or we still risk situation like this - I’ll send a note to the systemd ML for further clarification

**Comment by runcom** [comment]
looking at https://github.com/systemd/systemd/issues/10210 this also means, services relying on boot-complete should:

- add dd=no
- add conflicts and before shutdown.target

**Comment by lucab** [comment]
Even if the base OS may not (currently) ship that, custom Ignition configurations can install additional units required by `boot-complete.target`.
The same systemd docpage you linked explicitly mentions that usecases (`To support additional components that shall succeed before the boot is considered successful ...`).

**Comment by runcom** [comment]
> Even if the base OS may not (currently) ship that, custom Ignition configurations can install additional units required by `boot-complete.target`.
> The same systemd docpage you linked explicitly mentions that usecases (`To support additional components that shall succeed before the boot is considered successful ...`).

right, what I'm saying is if I ship an ignition right now that enables systemd-boot-check-no-failures, it'll introduce the cycle unless zincati et all aren't properly configured too - which is basically what I'm doing (just enable the before boot-complete service)

**Comment by runcom** [comment]
conversation going on here https://lists.freedesktop.org/archives/systemd-devel/2022-September/048330.html - the suggested fix would be to add After=multi-user.target to zincati et all

**Comment by lucab** [comment]
```
Note that systemd handling of .wants/ works like this:

1. add Wants= type dep
2. if no After=/Before= dep is set, then also add Before=

This means, that just adding an explicit After=multi-user.target to
your service means rule #2 won't take effect anymore.
```

I just learnt something new.
An explicit `After=` statement does indeed fix the dependency loop you observed. I'll go ahead and tweak Afterburn and Zincati accordingly.

**Comment by lucab** [comment]
PRs at https://github.com/coreos/afterburn/pull/802 and https://github.com/coreos/zincati/pull/831.

**Comment by lucab** [comment]
systemd docs update at https://github.com/systemd/systemd/pull/24741.

**Comment by dustymabe** [comment]
> PRs at [coreos/afterburn#802](https://github.com/coreos/afterburn/pull/802) and [coreos/zincati#831](https://github.com/coreos/zincati/pull/831).

Do any other services need to be updated?

**Comment by lucab** [comment]
I did look for other units queuing after `boot-complete.target` on latest FCOS `next`, and I didn't spot anything else to patch.

The only other cases are systemd own units (`systemd-bless-boot.service` and `systemd-machine-id-commit.service`) which already have `DefaultDependencies=no`.

**Comment by dustymabe** [comment]
I think that means we can close this?
